src/ZkFold/Symbolic/Base/Num.hs

```diff
@@ -0,0 +1,743 @@
+  => From Natural (Mod int n) where
+    from = UnsafeMod . residue @n . from
+
+instance (SemiIntegral int, KnownNat n)
+  => From Integer (Mod int n) where
+    from = UnsafeMod . from @Natural . Prelude.fromIntegral . residue @n
+
+instance (SemiIntegral int, Prime p)
+  => From Rational (Mod int p) where
+    from q = from (numerator q) / from (denominator q)
+
+instance Into (Mod int n) (Mod int n)
+
+instance Into Rational int
+  => Into Rational (Mod int n) where
+    to = to . fromMod
+
+instance Into Integer int
+  => Into Integer (Mod int n) where
+    to = to . fromMod
+
+instance SemiIntegral int
+  => Into Natural (Mod int n) where
+    to = Prelude.fromInteger . to
+
+instance (SemiIntegral int, KnownNat n)
+  => Scalar Natural (Mod int n) where
+    scale n z = from n * z
+    combine = combine . fmap (\(c,a) -> (from @_ @Natural c, a))
+instance (SemiIntegral int, KnownNat n)
+  => Scalar Integer (Mod int n) where
+    scale n z = from n * z
+    combine = combine . fmap (\(c,a) -> (from @_ @Integer c, a))
+instance (SemiIntegral int, Prime p)
+  => Scalar Rational (Mod int p) where
+    scale n z = from n * z
+    combine = combine . fmap (\(c,a) -> (from @_ @Rational c, a))
+instance (SemiIntegral int, KnownNat n)
+  => Scalar (Mod int n) (Mod int n) where
+    scale = (*)
+    combine = combineN . fmap (\(c,a) -> (to @Natural c, a))
+
+instance (SemiIntegral int, KnownNat n)
+  => Exponent Natural (Mod int n) where
+    exponent a q = exponent a (from @_ @(Mod int n) q)
+    evalMono = evalMono . fmap (\(a,q) -> (a,from @_ @(Mod int n) q))
+
+instance (SemiIntegral int, Prime p)
+  => Exponent Integer (Mod int p) where
+    exponent a q =
+      if q >= zero
+      then exponent a (from @_ @(Mod int p) q)
+      else one / exponent a (from @_ @(Mod int p) (negate q))
+    evalMono = evalMono . fmap absPow where
+      absPow (a,p) =
+        if p >= 0
+        then (a, from @_ @(Mod int p) p)
+        else (one / a, from (negate p))
+
+instance (SemiIntegral int, KnownNat n)
+  => Exponent (Mod int n) (Mod int n) where
+    exponent a q = from (to @Natural a ^ to @Natural q)
+    evalMono
+      = from
+      . evalMono
+      . fmap (\(a,q) -> (to @Natural a, to @Natural q))
+
+instance (SemiIntegral int, KnownNat n) => Discrete (Mod int n)
+instance (SemiIntegral int, KnownNat n) => Comparable (Mod int n)
+
+instance (KnownNat n, KnownNat (Log2 (n - 1) + 1))
+  => Finite (Mod int n) where
+    type Order (Mod int n) = n
+
+instance (SemiIntegral int, KnownNat n)
+  => FiniteChr (Mod int n) where
+    type Chr (Mod int n) = n
+
+instance (SemiIntegral int, Prime p, KnownNat (Log2 (p - 1) + 1), 3 <= p)
+  => Symbolic (Mod int p) (Mod int p)
+
+-- Scalar ------------------------------------------------------------------
+class (Semiring c, AdditiveMonoid a)
+  => Scalar c a where
+    scale :: c -> a -> a
+    default scale :: c ~ a => c -> a -> a
+    scale = (*)
+    combine :: [(c,a)] -> a
```

Comment:
A useful operation to be added to perform some operations faster via Fiat-Shamir trick, as explained to me by @echatav